File: STGGame/Assets/MainGame/Scripts/Package/MVC/MVCManager.cs
using System.Collections.Generic;
using XLibrary.Package;

namespace STGGame.MVC
{
    public class MVCManager : Singleton<MVCManager>
    {
        private Dictionary<string, Controller> m_controllerMaps = new Dictionary<string, Controller>();
        private Dictionary<string, Cache> m_cacheMaps = new Dictionary<string, Cache>();

        //Controller部分
        public T AddController<T>() where T : Controller, new()
        {
            T controller = new T();
            bool ret = controller.Initialize();
            if (ret)
            {
                string name = typeof(T).Name;
                m_controllerMaps[name] = controller;
            }
            return controller;
        }

        public void RemoveController<T>()
        {
            string name = typeof(T).Name;
            Controller controller = null;
            if (m_controllerMaps.TryGetValue(name,out controller))
            {
                controller.Dispose();
                m_controllerMaps.Remove(name);
            }
            
        }

        public T GetController<T>() where T : Controller, new()
        {
            string name = typeof(T).Name;
            Controller ctrl = null;
            if (!m_controllerMaps.TryGetValue(name,out ctrl))
            {
                ctrl = AddController<T>();
            }
            return ctrl as T;
        }

        //Cache部分
        public T AddCache<T>() where T : Cache, new()
        {
            string name = typeof(T).Name;
            T cache = new T();
            m_cacheMaps[name] = cache;
            
            return cache;
        }

        public void RemoveCache<T>()
        {
            Cache controller = null;
            string name = typeof(T).Name;
            if (m_cacheMaps.TryGetValue(name, out controller))
            {
                controller.Clear();
                m_cacheMaps.Remove(name);
            }
        }
        public T GetCache<T>() where T : Cache, new()
        {
            string name = typeof(T).Name;
            Cache cache = null;
            if (!m_cacheMaps.TryGetValue(name, out cache))
            {
                cache = AddCache<T>();
            }
            return cache as T;
        }

        public void ClearAllCaches()
        {
            foreach(var pair in m_cacheMaps)
            {
                pair.Value.Clear();
            }
        }

        public void ResetAllCaches()
        {
            m_cacheMaps.Clear();
        }
       

    }
}

File: Game/Script/Client/Base/Context/Patterns/MVC/MVCManager.lua
local M = class("MVCManager")

function M:ctor()
    self._ctrlDict = {}
    self._cacheDict = {}

    --注册一个轮询函数
    self._updateFunction = function ( ... )
        self:update(CSharp.Time.deltaTime)
    end

    --注册更新
    -- CSharp.MonoManagerIns:AddUpdateListener(self._updateFunction)
end

function M:addController(key,ctrl)
    if not ctrl then
        return false
    end
    if not key then
        return false
    end

    if not self._ctrlDict[key] then
        ctrl:initialize()
        self._ctrlDict[key] = ctrl
    end
    return true
end

function M:getController(key)
    return self._ctrlDict[key]
end

function M:removeController(key)
    local ctrl = self:getController(key)
    if ctrl then 
        ctrl:clear()
        self._ctrlDict[key] = nil
    end
end

function M:replaceController(key,ctrl)
    self:removeController(key)
    self:addController(key,ctrl)
end

function M:getControllerDict()
    return self._cacheDict
end

function M:addCache(key,cache)
    if not cache then
        return false
    end
    if not key then
        return false
    end

    if not self._cacheDict[key] then
        self._cacheDict[key] = cache 
    end
    return true
end

function M:removeCache(key)
    local cache = self:getCache(key)
    if cache then 
        cache:clear()
        self._cacheDict[key] = nil
    end
end

function M:replaceCache(key,cache)
    self:removeCache(key)
    self:addCache(key,cache)
end

function M:getCache(key)
    return self._cacheDict[key]
end

function M:getCacheDict()
    return self._cacheDict
end

function M:update(dt)

end


rawset(_G, "MVCManager", M)
MVCManager = M.new()